Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -84,28 +78,25 @@ public function findByIdentifier($identifier)
     }
 
     /**
-     * @param  AggregateRootInterface $aggregate
+     * @param AggregateRootInterface $aggregate
+     * @param int $expectedVersion
      * @return void
      */
-    public function save(AggregateRootInterface $aggregate)
+    final public function save(AggregateRootInterface $aggregate, int $expectedVersion = null)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
-        try {
-            $stream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            $stream = new EventStream();
+        if (!$aggregate instanceof EventSourcedAggregateRootInterface) {
+            throw new \InvalidArgumentException(sprintf('This repository can only save instances implementing EventSourcedAggregateRootInterface, given: "%s"', get_class($aggregate)), 1477226552);
         }
-
+        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
         $uncommittedEvents = $aggregate->pullUncommittedEvents();
-        $stream->addEvents(...$uncommittedEvents);
+        $version = $aggregate->getVersion() + 1;
+        $uncommittedEventsWithMetadata = array_map(function(EventInterface $event) use(&$version) {
+            return new EventWithMetadata($event, new EventMetadata([Metadata::VERSION => $version ++]));
+        }, $uncommittedEvents);
 
-        $this->eventStore->commit($streamName, $stream, function ($version) use ($uncommittedEvents) {
-            /** @var EventTransport $eventTransport */
-            foreach ($uncommittedEvents as $eventTransport) {
-                // @todo metadata enrichment must be done in external service, with some middleware support
-                $versionedMetadata = $eventTransport->getMetadata()->withProperty(Metadata::VERSION, $version);
-                $this->eventBus->handle($eventTransport->withMetadata($versionedMetadata));
-            }
-        });
+        if ($expectedVersion === null) {
+            $expectedVersion = $aggregate->getVersion();
```

Comment:
> didn't we decide to do optimistic concurrency check in the findBy*/loading of the aggregate?

Yes, but while implementing/testing this felt more consistent.
The observable behavior is mostly identical: I want to make sure that the aggregate hasn't been changed since I displayed it to the user.
The conflict resolution is also similar: Fetch missed events and compare them with the uncommitted aggregate events. If fine: call save again without the 2nd parameter